internal/parser/parser.go

```diff
@@ -0,0 +1,164 @@
+package parser
+
+import (
+	"errors"
+	"github.com/otrego/clamshell/internal/game"
+	"github.com/otrego/clamshell/internal/scanner"
+	"io"
+	"strings"
+)
+
+// the Parser uses a scanner to construct an Sgf object (the Game attribute)
+// By including the token and save attributes, we can "cache" one token
+// and "unscan" once
+type Parser struct {
+	Game    *game.Game
+	scanner *scanner.Scanner
+	token   *scanner.Token
+	save    bool
+}
+
+func FromString(s string) *Parser {
+	r := strings.NewReader(s)
+	return FromReader(r)
+}
+
+// create a new Parser struct, must initialize the Game attribute as well
+func FromReader(r io.Reader) *Parser {
+	game := game.NewGame()
+	return &Parser{
+		Game:    game,
+		scanner: scanner.New(r),
+	}
+}
+
+// scan one token using the scanner
+func (p *Parser) scan() *scanner.Token {
+	// the save bool is used to "rewind" once
+	if p.save {
+		p.save = false
+		return p.token
+	}
+	// otherwise just scan normally
+	p.token = p.scanner.Scan()
+	return p.token
+}
+
+// unscan lets the Parser know to use the saved token on the next scan
+// instead of scanning for a new token
+func (p *Parser) unscan() {
+	p.save = true
+}
+
+// scanSkipWhitespace is a wrapper around scan() that skips whitespace
+func (p *Parser) scanSkipWhitespace() *scanner.Token {
+	tok := p.scan()
+	if tok.Type == scanner.Whitespace {
+		tok = p.scan()
+	}
+	return tok
+}
+
+// Parses the root branch
+func (p *Parser) Parse() (*game.Game, error) {
+	if tok := p.scanSkipWhitespace(); tok.Type != scanner.LeftParen {
+		return nil, errors.New("Corrupted sgf: must start with '('")
+	}
+	g := game.NewGame()
+	err := p.parseBranch(g.Root)
+	if err != nil {
+		return nil, err
+	}
+	return g, nil
+}
+
+// parse, for example "PW[Player White]" into
+// field="PW"
+// value="Player White"
+func (p *Parser) parseFieldValue() (string, string, error) {
```

Comment:
How does this handle something like AW[pa][qa][nb][ob]?

Reply:
Not correctly, I'll tell you that :)

Reply:
Just had a shot at a fix